File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1
        
        while l < r:
            while l < r and not (s[l].isalpha() or s[l].isdigit()):
                l += 1
            while l < r and not (s[r].isalpha() or s[r].isdigit()):
                r -= 1
                
            if s[l].lower() != s[r].lower():
                return False
            
            l += 1
            r -= 1
        
        return True

File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int i = 0, j = s.length() - 1;
        
        while (i < j) {
            while(i < j && !isValidChar(s.charAt(i))) i++;
            while(i < j && !isValidChar(s.charAt(j))) j--;
            
            if (i >= j) break;
            
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) return false;
            
            i += 1;
            j -= 1;
        }
        
        return true;
    }
    
    boolean isValidChar(char c) {
        return Character.isAlphabetic(c) || Character.isDigit(c);
    }
}